### PR TITLE
field2* converters should not handle files

### LINKED ISSUE #558: Several possible problems within the ZPublisher Converters module
When working on #557 I spotted several possible problems within the Converters module.

The converters which inherit from `_unicode_converter` all fail under Python 3 and byte input.

Testcase:
```
def test_field2utext_with_binary_string(self):
    from ZPublisher.Converters import field2utext
    to_convert = b'abc'
    if PY2:
        expected = u'abc'
        self.assertEqual(field2utext(to_convert), expected)
    if PY3:
        expected = 'abc'
        self.assertEqual(field2utext(to_convert), expected)

Failure in test test_field2utext_with_binary_string (ZPublisher.tests.test_Converters.ConvertersTests)
Traceback (most recent call last):
  File "/usr/lib64/python3.6/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib64/python3.6/unittest/case.py", line 622, in run
    testMethod()
  File "/home/jugmac00/Projects/Zope/Zope/src/ZPublisher/tests/test_Converters.py", line 259, in test_field2utext_with_binary_string
    self.assertEqual(field2utext(to_convert), expected)
  File "/usr/lib64/python3.6/unittest/case.py", line 846, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib64/python3.6/unittest/case.py", line 1220, in assertMultiLineEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib64/python3.6/unittest/case.py", line 687, in fail
    raise self.failureException(msg)
AssertionError: "b'abc'" != 'abc'
- b'abc'
+ abc
```

This problem applies to field2utokens, field2utext, field2string. field2ulines on the other hand does not inherit from _unicode_converter and has to be examined separately.

I spotted https://github.com/zopefoundation/Zope/blob/5733da7d210d8fe322bfd03a01852bb02b529706/src/ZPublisher/Converters.py#L186 as the root problem for this behaviour, at least with Python 3.

----

Another possible problem is in https://github.com/zopefoundation/Zope/blob/5733da7d210d8fe322bfd03a01852bb02b529706/src/ZPublisher/Converters.py#L209 - as line breaks only get normalized when there is input of type byte.

I tried to fix these "problems" yesterday - seemingly successfully, but then suddenly https://github.com/zopefoundation/Zope/blob/5733da7d210d8fe322bfd03a01852bb02b529706/src/ZPublisher/tests/testHTTPRequest.py#L302 failed for the utext case.

----

Also, can anybody shed some light on what this "read" attribute is all about?
e.g. https://github.com/zopefoundation/Zope/blob/5733da7d210d8fe322bfd03a01852bb02b529706/src/ZPublisher/Converters.py#L221

Before spending more time on these issue, I'd like somebody to confirm that these are real issues.

### PR DESCRIPTION
**Pull request by jugmac00**
Until now field converters could be applied on files with unclear outcome.

The problems with trying to convert files to text (detect MIME type, guess charset...)
were discussed in length in https://github.com/zopefoundation/Zope/issues/558#issuecomment-485064104.

Handling files only makes sense for the `field2bytes` converter.

This fixes #558

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -20,6 +20,9 @@ https://github.com/zopefoundation/Zope/blob/4.x/CHANGES.rst
   ``rfc850_date``, and ``rfc1123_date`` which used to be in ``App.Common``
   keeping backwards-compatibility imports in place.
 
+- With the exception of ``field2bytes``, field converters should not handle
+  files (`#558 <https://github.com/zopefoundation/Zope/issues/558>`_)
```
Comment (icemac) [inline CHANGES.rst:24]: Without reading the diff it was not clear to me what has changed. `should not` is not really expressive. I'd expect something like "no longer try to handle file like objects special".

I think this is a even a breaking change and should be marked as such here. (Usually we use a headline for breaking changes.) It should be enough to increase the version number to 5.2.

File: src/ZPublisher/Converters.py [hunk 2]
```diff
@@ -23,20 +23,15 @@
 
 
 def field2string(v):
-    """Converts value to native strings.
-
-    So always to `str` no matter which Python version you are on.
-    """
-    if hasattr(v, 'read'):
-        return v.read()
-    elif isinstance(v, bytes):
```
Comment (icemac) [inline src/ZPublisher/Converters.py:32]: I am not sure if there should be a deprecation (maybe we could add it to Zope 4) or at least a warning (here in Zope 5).

**Review by icemac — CHANGES_REQUESTED** [review]
I general I like the idea, i just have some questions resp. suggestions.

File: CHANGES.rst (hunk 1 above)
Comment (jugmac00) [reply CHANGES.rst]: Thank you for the suggestions which I just applied.

File: src/ZPublisher/Converters.py (hunk 2 above)
Comment (jugmac00) [reply src/ZPublisher/Converters.py:32]: Well, that is up to you do decide.

I certainly agree it is a good idea to add a deprecation warning for Zope 4.

At the current state of this PR, a warning in Zope 5 is no longer reasonable, as the code path is gone.

If you insist on a deprecation warning in Zope 5, I'd need to create a new PR, and stall this one until Zope 6? I am not sure if this is necessary.

**Review by d-maurer — APPROVED** [review]
I do not think we need deprecation warnings in Zope 5: Zope 5 is Python 3 only and the file case is wrongly handled by the converters; thus, any user should already have been warned.

From my point of view, deprecation warnings are also not necessary in Zope 4: I assume that the "convert file" case is very rarely used.

**Review by icemac — APPROVED** [review]

File: src/ZPublisher/Converters.py (hunk 2 above)
Comment (icemac) [reply src/ZPublisher/Converters.py:32]: I am fine with the current solution.

**Comment by jugmac00** [comment]
@icemac and @d-maurer, thank you for taking the time for the review!

**Pull request merged**